File: 0463-island-perimeter/0463-island-perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

        def isValidMove(coord):
            if (
                coord[0] < 0 or
                coord[0] >= ROWS or
                coord[1] < 0 or
                coord[1] >= COLS or
                grid[coord[0]][coord[1]] == 0
            ):
                return False
            else:
                return True

        def getPerimiter(coord):
            visited = set([coord])
            perimiter = 0
            q = deque([coord])
            
            while q:
                curr = q.pop()
                for dc, dr in directions:
                    move = (curr[0] + dc, curr[1] + dr)

                    if isValidMove(move):
                        if move not in visited:
                            q.appendleft(move)
                    else:
                        perimiter += 1

                    visited.add(move)
            return perimiter
        
        for i in range(ROWS):
            for j in range(COLS):
                if grid[i][j] == 1:
                    return getPerimiter((i, j))

File: 0463-island-perimeter/0463-island-perimeter.go
func islandPerimeter(grid [][]int) int {

    prepend := func (s [][]int, element []int) [][]int {
        newSlice := make([][]int, len(s)+1)
        copy(newSlice[1:], s)
        newSlice[0] = element
        return newSlice
    }

    getPerimBFS := func(ROWS int, COLS int, start []int) int {
        type coord [2]int

        queued := map[coord]bool{}
        visited := map[coord]bool{
            coord{start[0], start[1]} : true,
        }

        directions := [4][2]int{
            {-1, 0}, // Up
            {1, 0},  // Down
            {0, -1}, // Left
            {0, 1},  // Right
        }

        q, result := [][]int{start}, 0

        for len(q) > 0 {
            curLoc := q[len(q) - 1]
            q = q[:len(q) - 1]
            fmt.Println("Location: ", curLoc, " res:", result)

            for _, dir := range directions {
                move := []int{
                    curLoc[0] + dir[0],
                    curLoc[1] + dir[1],
                }

                _, haveVisited := visited[coord{move[0], move[1]}]
                if haveVisited { continue }

                if ( move[0] < 0 ||
                     move[0] == ROWS ||
                     move[1] < 0 ||
                     move[1] == COLS ||
                     grid[move[0]][move[1]] == 0) { 
                     result++ 
                } else {
                    if _, haveQueued := queued[coord{move[0], move[1]}]; !haveQueued {
                        q = prepend(q, move)
                        queued[coord{move[0], move[1]}] = true
                    }
                }
                visited[coord{curLoc[0], curLoc[1]}] = true
            }
        }
        return result
    }

    for r := range grid {
        for c := range grid[r] {
            if grid[r][c] == 1 {
                return getPerimBFS(len(grid), len(grid[0]), []int{r, c})
            }
        }
    }
    return 0
}